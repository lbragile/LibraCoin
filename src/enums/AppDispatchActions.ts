export enum ACTIONS {
  ADD_VERIFIED_TRANS,
  UPDATE_VERIFIED_TRANS,
  UPDATE_SELECTED_TRANS,
  UPDATE_USERS,
  SET_MAIN_USER,
  ADD_BLOCK,
  UPDATE_BLOCK
}
