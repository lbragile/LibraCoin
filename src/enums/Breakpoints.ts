// https://www.w3schools.com/css/css_rwd_mediaqueries.asp

export enum SIZES {
  xs = "600px",
  sm = "600px",
  md = "768px",
  lg = "992px",
  xl = "1200px"
}

export enum DEVICES {
  xs = "max-width: " + SIZES.xs,
  sm = "min-width: " + SIZES.sm,
  md = "min-width: " + SIZES.md,
  lg = "min-width: " + SIZES.lg,
  xl = "min-width: " + SIZES.xl
}
