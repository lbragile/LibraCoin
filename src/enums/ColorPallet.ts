export enum COLORS {
  DEFAULT_BACKGROUND = "#fff",

  DARK_BACKGROUND = "#1f1f1f",

  VALID_BACKGROUND = "#b4ff9a",
  INVALID_BACKGROUND = "#ff9a9a",
  VALID_SOLUTION = "#008000",
  INVALID_SOLUTION = "#f00",

  NAVBAR_BACKGROUND = "#eee",
  NAV_LINK_UNDERSCORE = "#8ac0ff",
  NAV_TEXT = "#000",

  SELECTED_TRANS_BACKGROUND = "#dfd",
  SELECTED_TRANS_BORDER = "#090",

  BLOCK_TRANS_BACKGROUND = "#c7e4ff",

  ROOT_CELL = "#4f9249",
  MIDDLE_CELLS = "#cf8b31",
  DATA_CELLS = "#7d95ff",
  PREVIEW_LINES = "#777"
}
