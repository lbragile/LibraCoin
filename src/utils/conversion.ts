import { TextEncoder } from "text-encoding";

export function bufferToHex(hashBuffer: ArrayBuffer): string {
  const hashArray = Array.from(new Uint8Array(hashBuffer)); // convert buffer to byte array
  const hashHex = hashArray.map((b) => b.toString(16).padStart(2, "0")).join(""); // convert bytes to hex string
  return hashHex;
}

export async function CryptoKeyToHex(format: string, key: CryptoKey): Promise<string> {
  const buf = (await window.crypto.subtle.exportKey(format, key)) as ArrayBuffer;
  return bufferToHex(buf);
}

export async function digestMessage(message: string): Promise<string> {
  const msgUint8 = new TextEncoder().encode(message); // encode as (utf-8) Uint8Array
  const hashBuffer = await window.crypto.subtle.digest("SHA-256", msgUint8); // hash the message
  return bufferToHex(hashBuffer);
}

export function randomHash(len: number): string {
  return bufferToHex(window.crypto.getRandomValues(new Uint32Array(len)));
}
