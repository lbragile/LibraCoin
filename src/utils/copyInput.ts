export function copyKey(
  e: React.FocusEvent<HTMLTextAreaElement | HTMLInputElement>,
  setCopied: (arg: boolean[]) => void,
  type?: "public" | "private"
): void {
  e.target.select();
  e.target.setSelectionRange(0, 1e6);
  document.execCommand("copy");

  // wallet page, copying keys
  if (type) {
    const isPublic = type === "public";
    setCopied([isPublic, !isPublic && !e.target.value.includes("â—¦")]);
  } else {
    setCopied([true]);
  }
}
