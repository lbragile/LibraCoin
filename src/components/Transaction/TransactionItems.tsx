import React, { useContext } from "react";
import { Form, InputGroup } from "react-bootstrap";
import { AppContext } from "../../context/AppContext";
import { ACTIONS } from "../../enums/AppDispatchActions";
import { IAction, IState, ITransaction } from "../../typings/AppTypes";

import "./Transaction.css";

export default function TransactionItems(): JSX.Element {
  const { state, dispatch } = useContext(AppContext) as { state: IState; dispatch: React.Dispatch<IAction> };

  function selectTransaction(transaction: ITransaction): void {
    let selectedTrans: ITransaction[] = JSON.parse(JSON.stringify(state)).selectedTrans;
    const signatures = selectedTrans.map((x) => x.signature);
    const included = signatures.includes(transaction.signature);

    if (selectedTrans.length < 4 || included) {
      // if just selected - push onto stack, else remove it
      // adjust backgrounds according to selection/deselection
      if (!included) {
        selectedTrans.push(transaction);
      } else {
        selectedTrans = selectedTrans.filter((x) => x.signature !== transaction.signature);
      }

      dispatch({ type: ACTIONS.UPDATE_SELECTED_TRANS, payload: { selectedTrans } });
    } else {
      alert("You can mine at most 4 transactions at a time!");
    }
  }

  return (
    <div className="container-fluid">
      <h3 className="font-weight-bold">Verified Transactions</h3>
      <div className="trans-list row flex-nowrap overflow-auto bg-dark mx-1 px-2 rounded">
        {state.verifiedTrans.map((transaction: ITransaction, i: number) => {
          return (
            <div
              className={
                "trans-item " +
                (state.selectedTrans.map((x) => x.signature).includes(transaction.signature) ? "selected" : "not-selected") // prettier-ignore
              }
              onClick={() => selectTransaction(transaction)}
              key={`verifiedTrans${i}`}
            >
              <Form.Group className="mb-2 text-center">
                <Form.Control className="text-truncate" type="text" defaultValue={transaction.from} readOnly />
                <h3 className="my-0">â†“</h3>
                <Form.Control className="text-truncate" type="text" defaultValue={transaction.to} readOnly />
              </Form.Group>

              <InputGroup className="mb-2">
                <InputGroup.Prepend>
                  <InputGroup.Text>Msg</InputGroup.Text>
                </InputGroup.Prepend>
                <Form.Control as="textarea" defaultValue={transaction.message} readOnly />
              </InputGroup>

              <InputGroup className="mb-2">
                <Form.Control type="number" defaultValue={transaction.amount} disabled />
                <InputGroup.Append>
                  <InputGroup.Text>LC</InputGroup.Text>
                </InputGroup.Append>
              </InputGroup>

              <InputGroup>
                <InputGroup.Prepend>
                  <InputGroup.Text>Sig</InputGroup.Text>
                </InputGroup.Prepend>
                <Form.Control className="text-truncate" type="text" defaultValue={transaction.signature} readOnly />
              </InputGroup>
            </div>
          );
        })}
      </div>
    </div>
  );
}
