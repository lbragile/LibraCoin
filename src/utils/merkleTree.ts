import { Tree } from "./Tree";
import { ITransaction } from "../typings/AppTypes";
import { digestMessage } from "./conversion";

export async function calculateMerkleTreeFormation(
  verifiedTrans: ITransaction[],
  selectedTrans: ITransaction[],
  setMerkleTree?: (arg: string[][]) => void
): Promise<string> {
  let tree = [[""]];
  if (selectedTrans.length > 0) {
    // need to make sure node's in tree appear in same order as in the verified transaction pane, regardless of selection order
    const verifiedSignatures = verifiedTrans.map((trans) => trans.signature);
    const selectedSignatures = selectedTrans.map((trans) => trans.signature);
    let signatures = verifiedSignatures.filter((sig) => selectedSignatures.includes(sig));
    tree = [signatures];

    while (signatures.length !== 1) {
      const hashArr = [] as string[];
      for (let i = 0; i < signatures.length; i += 2) {
        const hash = signatures[i + 1] ? await digestMessage(signatures[i] + signatures[i + 1]) : signatures[i];
        hashArr.push(hash);
      }

      signatures = hashArr;
      tree.push(hashArr);
    }
  }

  if (setMerkleTree) {
    setMerkleTree(tree);
  }

  return getMerkleRoot(tree);
}

export function getMerkleRoot(tree: string[][]): string {
  return tree[tree.length - 1][0];
}

export function flattenTree(tree: string[][]): string[] {
  let flatTree = [] as string[];
  // need to reverse each tree row so that printing produces the right order!
  for (let i = 0; i < tree.length; i++) {
    flatTree = flatTree.concat(tree[i].reverse());
  }

  return flatTree;
}

export function drawTreeDiagramOnCanvas(
  merkleTree: string[][],
  canvas: HTMLCanvasElement | null,
  transactions: ITransaction[]
): void {
  if (canvas) {
    const canvasTree = new Tree(canvas, transactions);
    canvasTree.clear();
    const flatTree = flattenTree(merkleTree);
    for (let i = flatTree.length - 1; i >= 0; i--) {
      canvasTree.addNode(flatTree[i]);
    }
    canvasTree.drawTree();
  }
}
