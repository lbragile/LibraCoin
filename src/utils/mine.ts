import { digestMessage, randomHash } from "./conversion";

async function createTarget(numZeros: number): Promise<string> {
  const targetHash = await digestMessage(randomHash(20));

  // replace leading bits with zeros
  const re = new RegExp(`^.{0,${numZeros}}`, "g");
  const zerosStr = Array(numZeros).fill("0").join("");

  return targetHash.replace(re, zerosStr);
}

export async function mine(
  nonce: number,
  setHeader: (arg: number) => void,
  setTarget: (arg: string) => void,
  setSolution: (arg: string) => void,
  setIsValid: (arg: boolean) => void
): Promise<void> {
  // make target
  const numZeros = Math.round(Math.random()) + 2;
  const targetHash = await createTarget(numZeros);
  setTarget(targetHash);

  // mine
  let candidateSolution = "";
  let header = nonce; // re-assigning for clarity
  while (header <= Number.MAX_SAFE_INTEGER) {
    candidateSolution = await digestMessage(header.toString());
    setSolution(candidateSolution);
    setHeader(header++);

    const leadingBits = candidateSolution.substr(0, numZeros).split("");
    if (leadingBits.every((bit) => bit === "0")) {
      break;
    }
  }

  setIsValid(candidateSolution <= targetHash);
}
