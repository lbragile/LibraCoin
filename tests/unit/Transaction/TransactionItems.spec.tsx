/**
 * @group unit
 */

import React from "react";
import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import TransactionItems from "../../../src/components/Transaction/TransactionItems";
import { ACTIONS } from "../../../src/enums/AppDispatchActions";

import * as TreeUtils from "../../../src/utils/merkleTree";
import { customRender } from "../../utils/testUtils";

const { initialState } = global;

it("renders correctly", () => {
  const { asFragment } = customRender(<TransactionItems />);

  const transactions = screen.getAllByRole("form", { name: /Transaction Information/i });
  const fromFields = screen.getAllByRole("textbox", { name: /Transaction From/i });
  const toFields = screen.getAllByRole("textbox", { name: /Transaction To/i });
  const msgFields = screen.getAllByRole("textbox", { name: /Transaction Message/i });
  const balances = screen.getAllByRole("spinbutton", { name: /Transaction Amount/i });
  const sigFields = screen.getAllByRole("textbox", { name: /Transaction Signature/i });

  expect(screen.getByRole("heading", { name: /Title/i, level: 3 })).toHaveTextContent("Verified Transactions");

  transactions.forEach((transaction, i) => expect(transaction).toHaveFormValues({ ...initialState.verifiedTrans[i] }));

  [...fromFields, ...toFields, ...msgFields, ...sigFields].forEach((field) => {
    expect(field).toHaveAttribute("readOnly");
    expect(field).not.toBeRequired();
  });

  balances.forEach((balance) => {
    expect(balance).toBeDisabled();
    expect(balance).not.toBeRequired();
  });

  const numTrans = initialState.verifiedTrans.length;
  expect(transactions.length).toEqual(numTrans);
  expect(fromFields.length).toEqual(numTrans);
  expect(toFields.length).toEqual(numTrans);
  expect(msgFields.length).toEqual(numTrans);
  expect(balances.length).toEqual(numTrans);
  expect(sigFields.length).toEqual(numTrans);

  expect(asFragment()).toMatchSnapshot();
});

describe("Select Transaction", () => {
  const dispatchMock = jest.fn();

  describe("Less than 4 already selected", () => {
    beforeEach(() => customRender(<TransactionItems />, { dispatchMock }));
    afterEach(() => jest.clearAllMocks());

    it("adds selected transaction if not already selected", async () => {
      const transactions = screen.getAllByRole("form", { name: /Transaction Information/i });
      const fromFields = screen.getAllByRole("textbox", { name: /Transaction From/i }) as HTMLInputElement[];
      const toFields = screen.getAllByRole("textbox", { name: /Transaction To/i }) as HTMLInputElement[];
      const msgFields = screen.getAllByRole("textbox", { name: /Transaction Message/i }) as HTMLTextAreaElement[];
      const balances = screen.getAllByRole("spinbutton", { name: /Transaction Amount/i }) as HTMLInputElement[];
      const sigFields = screen.getAllByRole("textbox", { name: /Transaction Signature/i }) as HTMLInputElement[];

      const newSelected = {
        amount: +balances[0].value,
        from: fromFields[0].value,
        msg: msgFields[0].value,
        signature: sigFields[0].value,
        to: toFields[0].value
      };

      // new merkleRoot will be just the transaction signature since only 1 selected
      jest.spyOn(TreeUtils, "calculateMerkleTreeFormation").mockResolvedValueOnce([[sigFields[0].value]]);
      const newPreview = { ...initialState.preview, merkleRoot: sigFields[0].value, valid: false };

      const selectedTrans = [...initialState.selectedTrans, newSelected]; // expected

      userEvent.click(transactions[0]);

      await waitFor(() => expect(dispatchMock).toHaveBeenCalledTimes(2));

      expect(dispatchMock).toHaveBeenNthCalledWith(1, {
        type: ACTIONS.UPDATE_SELECTED_TRANS,
        payload: { selectedTrans }
      });

      expect(dispatchMock).toHaveBeenNthCalledWith(2, {
        type: ACTIONS.UPDATE_PREVIEW,
        payload: { preview: newPreview }
      });
    });

    it("removes selected transaction from the list if already selected", async () => {
      const transactions = screen.getAllByRole("form", { name: /Transaction Information/i });

      jest.spyOn(TreeUtils, "calculateMerkleTreeFormation").mockResolvedValueOnce([[""]]);
      const newPreview = { ...initialState.preview, merkleRoot: "", valid: false };

      userEvent.click(transactions[1]);

      await waitFor(() => expect(dispatchMock).toHaveBeenCalledTimes(2));

      expect(dispatchMock).toHaveBeenNthCalledWith(1, {
        type: ACTIONS.UPDATE_SELECTED_TRANS,
        payload: { selectedTrans: [] }
      });

      expect(dispatchMock).toHaveBeenNthCalledWith(2, {
        type: ACTIONS.UPDATE_PREVIEW,
        payload: { preview: newPreview }
      });
    });
  });

  describe("4 or more already selected", () => {
    const dispatchMock = jest.fn();
    const ogState = JSON.parse(JSON.stringify(initialState));
    ogState.selectedTrans.push(...ogState.verifiedTrans.slice(2));

    beforeEach(() => customRender(<TransactionItems />, { stateMock: ogState, dispatchMock }));
    afterEach(() => jest.clearAllMocks());

    it("alerts", () => {
      const alertSpy = jest.spyOn(window, "alert").mockImplementation(() => undefined);

      const transactions = screen.getAllByRole("form", { name: /Transaction Information/i });
      userEvent.click(transactions[0]);

      expect(alertSpy).toHaveBeenCalledTimes(1);
      expect(alertSpy).toHaveBeenCalledWith("You can mine at most 4 transactions at a time!");
    });

    it("removes the selectedTrans from the list without alerting if the clicked transaction matches it", async () => {
      const alertSpy = jest.spyOn(window, "alert").mockImplementation(() => undefined);

      const transactions = screen.getAllByRole("form", { name: /Transaction Information/i });
      const selectedTrans = ogState.selectedTrans.slice(0, ogState.selectedTrans.length - 1); // expected

      jest.spyOn(TreeUtils, "calculateMerkleTreeFormation").mockResolvedValueOnce([["randomRoot"]]);
      const newPreview = { ...initialState.preview, merkleRoot: "randomRoot", valid: false };

      userEvent.click(transactions[transactions.length - 1]);

      await waitFor(() => expect(dispatchMock).toHaveBeenCalledTimes(2));

      expect(dispatchMock).toHaveBeenNthCalledWith(1, {
        type: ACTIONS.UPDATE_SELECTED_TRANS,
        payload: { selectedTrans }
      });

      expect(dispatchMock).toHaveBeenNthCalledWith(2, {
        type: ACTIONS.UPDATE_PREVIEW,
        payload: { preview: newPreview }
      });

      expect(alertSpy).not.toHaveBeenCalled();
    });
  });
});
