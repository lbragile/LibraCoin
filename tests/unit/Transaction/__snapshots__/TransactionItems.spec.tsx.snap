// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container-fluid"
  >
    <h3
      aria-label="Title"
      class="font-weight-bold"
    >
      Verified Transactions
    </h3>
    <div
      class="trans-list row flex-nowrap overflow-auto bg-dark mx-1 px-2 rounded"
    >
      <form
        aria-label="Transaction Information"
        class="trans-item not-selected"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="B"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            First Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="1.23"
          />
          <div
            class="input-group-append"
          >
            <span
              class="input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AB1.23"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="trans-item selected"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="C"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Second Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="456.78"
          />
          <div
            class="input-group-append"
          >
            <span
              class="input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AC456.78"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="trans-item not-selected"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="D"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Third Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="999.99"
          />
          <div
            class="input-group-append"
          >
            <span
              class="input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AD999.99"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="trans-item not-selected"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="E"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Forth Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="987.65"
          />
          <div
            class="input-group-append"
          >
            <span
              class="input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AE987.65"
          />
        </div>
      </form>
      <form
        aria-label="Transaction Information"
        class="trans-item not-selected"
      >
        <div
          class="mb-2 text-center form-group"
        >
          <input
            aria-label="Transaction From"
            class="text-truncate form-control"
            name="from"
            readonly=""
            type="text"
            value="F"
          />
          <h3
            class="my-0"
          >
            ↓
          </h3>
          <input
            aria-label="Transaction To"
            class="text-truncate form-control"
            name="to"
            readonly=""
            type="text"
            value="A"
          />
        </div>
        <div
          class="mb-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Msg
            </span>
          </div>
          <textarea
            aria-label="Transaction Message"
            class="form-control"
            name="msg"
            readonly=""
          >
            Fifth Transaction
          </textarea>
        </div>
        <div
          class="mb-2 input-group"
        >
          <input
            aria-label="Transaction Amount"
            class="form-control"
            disabled=""
            name="amount"
            type="number"
            value="1.01"
          />
          <div
            class="input-group-append"
          >
            <span
              class="input-group-text"
            >
              LC
            </span>
          </div>
        </div>
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Sig
            </span>
          </div>
          <input
            aria-label="Transaction Signature"
            class="text-truncate form-control"
            name="signature"
            readonly=""
            type="text"
            value="AF1.01"
          />
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
